src/ucp/core/ucp_ep.c

```diff
@@ -206,45 +210,98 @@ ucs_status_t ucp_ep_create_base(ucp_worker_h worker, const char *peer_name,
 #if UCS_ENABLE_ASSERT
     ucp_ep_ext_control(ep)->ka_last_round = 0;
 #endif
-
     UCS_STATIC_ASSERT(sizeof(ucp_ep_ext_gen(ep)->ep_match) >=
                       sizeof(ucp_ep_ext_gen(ep)->flush_state));
     memset(&ucp_ep_ext_gen(ep)->ep_match, 0,
            sizeof(ucp_ep_ext_gen(ep)->ep_match));
 
     ucs_hlist_head_init(&ucp_ep_ext_gen(ep)->proto_reqs);
-    ucp_stream_ep_init(ep);
-    ucp_am_ep_init(ep);
 
     for (lane = 0; lane < UCP_MAX_LANES; ++lane) {
         ep->uct_eps[lane] = NULL;
     }
-
 #if ENABLE_DEBUG_DATA
     ucs_snprintf_zero(ep->peer_name, UCP_WORKER_ADDRESS_NAME_MAX, "%s",
                       peer_name);
 #endif
-
     /* Create statistics */
     status = UCS_STATS_NODE_ALLOC(&ep->stats, &ucp_ep_stats_class,
                                   worker->stats, "-%p", ep);
     if (status != UCS_OK) {
         goto err_free_ep_control_ext;
     }
 
+    return ep;
+
+err_free_ep_control_ext:
+    ucs_free(ucp_ep_ext_control(ep));
+err_free_ep:
+    ucs_strided_alloc_put(&worker->ep_alloc, ep);
+err:
+    return NULL;
+}
+
+static int ucp_ep_shall_use_indirect_id(ucp_context_h context,
+                                        unsigned ep_init_flags)
+{
+    return !(ep_init_flags & UCP_EP_INIT_FLAG_INTERNAL) &&
+           ((context->config.ext.proto_indirect_id == UCS_CONFIG_ON) ||
+            ((context->config.ext.proto_indirect_id == UCS_CONFIG_AUTO) &&
+             (ep_init_flags & UCP_EP_INIT_ERR_MODE_PEER_FAILURE)));
+}
+
+ucs_status_t ucp_ep_create_base(ucp_worker_h worker, unsigned ep_init_flags,
+                                const char *peer_name, const char *message,
+                                ucp_ep_h *ep_p)
+{
+    ucs_status_t status;
+    ucp_ep_h ep;
+
+    ep = ucp_ep_allocate(worker, peer_name);
+    if (ep == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err;
```

Comment:
```suggestion
        return UCS_ERR_NO_MEMORY;
```
and remove line 305.

Reply:
I wanted to have a single exit point in the function.
e.g. if we will add more logic that requires some reverse action, it will be easy to just add a new `goto` label which invokes this reverse action.

Reply:
I see. The comment is relevant only if the function has the final form.